libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/builder/NavRouteOptionsBuilder.kt

```diff
@@ -0,0 +1,362 @@
+package com.mapbox.navigation.core.routeoptions.builder
+
+import android.content.Context
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.Bearing
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.extensions.applyDefaultNavigationOptions
+import com.mapbox.navigation.base.extensions.applyLanguageAndVoiceUnitOptions
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface NoWaypointsOptionsBuilder {
+    fun fromStartLocation(
+        point: Point,
+        bearing: Double? = null,
+        zLevel: Int?
+    ): WaypointsInProgressBuilder
+
+    fun fromCurrentLocation(): WaypointsInProgressBuilder
+}
```

Comment:
> Builders are meant to be flexible and allow developers to call any setter method in any order before calling build()

The builder pattern is just a pattern, i.e. I think it meant to be as the majority of developers think it is 🙂 
And I've seen recently a description [of a step builder pattern in AutoValue docs](https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#step), which I think is quite popular. So I can't agree with "Builders are meant to be flexible" for now.

I agree that the "step builder" pattern doesn't allow re-use the builder between calls, because it changes the interface on each call. 

I see a tradeoff here. From one side of the tradeoof is flexibility, from the other side is safety. The "step builder" pattern ensures that all the required fields are filled in. With current approach we get safety sacrificing flexibility. 

@tomaszrybakiewicz, what do you think about it? Do you think it better to pick flexibility over safety?

Reply:
Both are not mutually exclusive. You can offer a flexible API to developers and enforce its safety on **runtime** via assertions/preconditions. Lots of libraries does that. Take [firebase-android-sdk](https://github.com/firebase/firebase-android-sdk/search?q=checkArgument) for example. Even though they use `@NonNull` annotations and explicitly mention argument non nullability requirement in documentation, they still use preconditions to validate assumptions and enforce correct api usage.

Reply:
This discussion is very similar to the [discussion above](https://github.com/mapbox/mapbox-navigation-android/pull/5427#discussion_r830039141).
I will try different API formats before making this API not experimental. But I think results of first try is good enough to be merged.